projects/transformers/callbacks/metrics.py

```diff
@@ -25,104 +25,103 @@
-
-            # guarantee that everything stayed sparse,
-            # up to specified tolerance
-            if (self.sparsity_tolerance < 1) and len(self.eval_metrics["sparsity"]) > 1:
-                sparse_diff = self.eval_metrics["sparsity"][0] - self.eval_metrics["sparsity"][-1]  # noqa
-                if abs(sparse_diff) > self.sparsity_tolerance:
-                    logging.warn(
-                        "Model sparsity fluctuated beyond acceptable range."
-                        f"Current sparsity level: {self.eval_metrics['sparsity'][-1]}"
-                    )
-
-            # track learning rate
-            # get_last_lr() returns lr for each parameter group. For now,
-            # assume lrs are the same for all and just track one.
-            if kwargs["lr_scheduler"] is not None:
-                last_lr = kwargs["lr_scheduler"].get_last_lr()
-                self.eval_metrics["lr"].append(last_lr[0])
-
-            # TODO
-            # Possibly update train_results
-            # Possibly wandb logging
+        """Update eval metrics and possibly step counter, sparsity, and lr"""
 
-    def on_train_begin(self, args, state, control, **kwargs):
-        """
-        Stop from accumulating results accross multiple runs
-        """
-        self.__init__()
+        # track performance metrics
+        for key in metrics.keys():
+            if key not in self.eval_metrics:
+                self.eval_metrics[key] = [metrics[key]]
+            else:
+                self.eval_metrics[key].append(metrics[key])
 
-    def check_for_mnli(self, metrics):
+        self.call_counter += 1
+        do_remaining_updates = self.maybe_update()
 
-        # Using a workaround for mnli, and only evaluating mnli-mm at the very
-        # end of training. Therefore don't update anything
-        is_mnli_mm = "mm_accuracy" in metrics.keys()
+        # Update steps, sparsity, learning rate
+        if do_remaining_updates:
+            self.update_auxillary_metrics(args, **kwargs)
 
-        # flake8 wants this to be a def, not a lambda
-        def ksplit(k):
-            return k.split("_", 1)[1] if "_" in k else k
+    def maybe_update(self):
+        """
+        If just one eval set
+            update step counter, sparsity, etc.
+        Else
+            update only if you finished cycling through all eval sets
+        """
+        do_remaining_updates = False
+        if self.n_eval_sets > 1:
+            if self.call_counter % self.n_eval_sets == 0:
+                do_remaining_updates = True
+        else:
+            do_remaining_updates = True
 
-        if is_mnli_mm:
-            self.mm_metrics = {ksplit(k): v for k, v in metrics.items()}
+        return do_remaining_updates
 
-        return is_mnli_mm
+    def on_train_begin(self, args, state, control, **kwargs):
+        """
+        Stop from accumulating results accross multiple runs
+        """
+        self.__init__()
```

Comment:
Is this the same thing as the original init, shouldn't it take arguments? Is there another way of reinitializing result tracking without calling init?

Reply:
Good catch. That absolutely should have arguments. RE not calling init, the alternative would be to manually reset variables like self.step_counter and self.eval_metrics. But that's all the init does anyway, so I chose to call it.